pkg/pgmodel/ingestor/trace/trace_batcher.go

```diff
@@ -0,0 +1,248 @@
+	if defaultBatchWorkers == 0 {
+		// in case there <=1 CPU
+		defaultBatchWorkers = 1
+	}
+}
+
+// We batch individual InsertTracesReq
+type Batch struct {
+	traces    ptrace.Traces
+	spanCount int
+	reqStatus []chan error
+	ctx       context.Context
+	maxSize   int
+}
+
+func NewBatch(maxBatchSize int) *Batch {
+	return &Batch{
+		traces:    ptrace.NewTraces(),
+		spanCount: 0,
+		reqStatus: make([]chan error, 0, maxBatchSize),
+		ctx:       context.Background(),
+		maxSize:   maxBatchSize, // this is not hard limit
+	}
+}
+
+func (tb *Batch) add(in InsertTracesReq) {
+	_, addReqSpan := tracer.Default().Start(tb.ctx, "add-trace-req-to-batch",
+		trace.WithLinks(trace.Link{SpanContext: in.spanCtx}))
+	defer addReqSpan.End()
+	inSpans := in.payload.SpanCount()
+	if inSpans == 0 {
+		return
+	}
+	in.payload.ResourceSpans().MoveAndAppendTo(tb.traces.ResourceSpans())
+	tb.spanCount += inSpans
+	tb.reqStatus = append(tb.reqStatus, in.err)
+}
+
+func (tb *Batch) isFull() bool {
+	return tb.spanCount >= tb.maxSize
+}
+
+func (tb *Batch) isEmpty() bool {
+	return tb.spanCount == 0
+}
+
+// Batcher batches trace requests and sends batches to batch writer
+// This is done to achieve better ingest performance especially b/c
+// Jaeger collector sends traces one by one
+type Batcher struct {
+	in              chan InsertTracesReq
+	stop            chan struct{}
+	batchWriter     *batchWriter
+	once            sync.Once
+	bufferedBatches chan Batch
+	wg              sync.WaitGroup
+	config          BatcherConfig
+}
+
+type BatcherConfig struct {
+	Batchers     int
+	Writers      int
+	MaxBatchSize int
+	BatchTimeout time.Duration
+}
+
+type WriteTraces func(context.Context, ptrace.Traces) error
+
+func NewBatcher(config BatcherConfig, writer Writer) *Batcher {
+	if config.Batchers == 0 {
+		config.Batchers = defaultBatchWorkers
+	}
+	if config.MaxBatchSize == 0 {
+		config.MaxBatchSize = DefaultBatchSize
+	}
+	if config.BatchTimeout == 0 {
+		config.BatchTimeout = DefaultBatchTimeout
+	}
+	if config.Batchers < 1 || config.Writers < 1 {
+		panic("number of batchers and writeres must be greater then zero")
+	}
+	bufferedBatches := make(chan Batch, defaultMaxBufferedBatches)
+	return &Batcher{
+		batchWriter:     newBatchWriter(config.Writers, writer, bufferedBatches),
+		in:              make(chan InsertTracesReq, defaultReqBufferSize),
+		stop:            make(chan struct{}, 1),
+		bufferedBatches: bufferedBatches,
+		config:          config,
+	}
+}
+
+func (b *Batcher) Run() {
+	for i := 0; i < b.config.Batchers; i++ {
+		b.wg.Add(1)
+		go func() {
+			defer b.wg.Done()
+			b.batch()
+		}()
+	}
+	b.batchWriter.run()
+}
+
+func (b *Batcher) batch() {
+	_, batcherSpan := tracer.Default().Start(context.Background(), "create-batches")
+	ticker := time.NewTicker(b.config.BatchTimeout)
+	batch := NewBatch(b.config.MaxBatchSize)
+	batcherSpan.AddEvent("New Batch")
+	flushBatch := func(batch *Batch) *Batch {
+		batchCp := *batch
+		b.bufferedBatches <- batchCp
```

Comment:
Why not send in the channel by reference? that is more performant than sending direct values.
```suggestion
		b.bufferedBatches <- batch or batchCp
```
Also, I am unable to find how we are reusing the batch for using a copy.

Reply:
I believe this is more about semantics, not performance. Semantically sending copy more clearly communicates that batch is immutable. I've run quick benchmark for `pointer vs struct` and they are almost identical:
```
BenchmarkSendBatchStruct-10       153936          8596 ns/op       41552 B/op          4 allocs/op
BenchmarkSendBatchPointer-10      143161          8936 ns/op       41619 B/op          6 allocs/op
```
I do agree that when sending large structs it's better to use pointers however not the case here...

Reply:
Can you share the batch size in this benchmark please? Ah, nvm, `ptrace.Traces` in itself is a slice.

Reply:
```suggestion
		b.bufferedBatches <- *batch
```
If we want to use a copy, then how about this? This will save the stack allocation, which would otherwise be the size of the batch.

Reply:
About the benchmark, its right that difference wouldn't be huge, since `ptrace.Traces` is internally a slice :). Hence, we can ignore the pointer stuff, but see above comment on stack size.